src/v/kafka/server/usage_manager.cc

```diff
@@ -0,0 +1,386 @@
+    /// called after restart to determine which bucket is the 'current' bucket
+    auto last_window_delta = std::chrono::seconds(0);
+    _current_window = 0;
+    if (!buckets.empty()) {
+        std::optional<size_t> open_index;
+        for (size_t i = 0; i < buckets.size(); ++i) {
+            /// There will always be exactly 1 open_window in the result set
+            if (buckets.at(i).is_open()) {
+                vassert(!open_index, "Data serialization was incorrect");
+                open_index = i;
+                break;
+            }
+        }
+        vassert(open_index, "Data serialization was incorrect");
+        /// Optimization to begin picking up if wall time is within
+        /// window interval
+        const auto begin = std::chrono::seconds(buckets.at(*open_index).begin);
+        const auto now = std::chrono::seconds(epoch_time_secs());
+        vassert(begin < now, "Clock or data serialization error");
+        const auto delta = now - begin;
+        if (delta < _usage_window_width_interval) {
+            /// Continue from where the program left off
+            _current_window = *open_index;
+            last_window_delta = delta;
+        } else {
+            /// Close window and open a new one
+            buckets.at(*open_index).end = epoch_time_secs();
+            _current_window = (*open_index + 1) % buckets.size();
+            buckets.at(_current_window).reset(ss::lowres_system_clock::now());
+        }
+    }
+    _buckets = std::move(buckets);
+    return last_window_delta;
+}
+
+/// Operated on only by shard-0 this prevents execution of background work
+/// spawned by the configuration property update callbacks after stop has been
+/// called on the usage_manager
+static thread_local ss::gate bg_future_gate;
+
+usage_manager::usage_manager(ss::sharded<storage::api>& storage)
+  : _usage_enabled(config::shard_local_cfg().enable_usage.bind())
+  , _usage_num_windows(config::shard_local_cfg().usage_num_windows.bind())
+  , _usage_window_width_interval(
+      config::shard_local_cfg().usage_window_width_interval_sec.bind())
+  , _usage_disk_persistance_interval(
+      config::shard_local_cfg().usage_disk_persistance_interval_sec.bind())
+  , _storage(storage) {
+    _usage_enabled.watch([this] { (void)reset(); });
+    _usage_num_windows.watch([this] { (void)reset(); });
+    _usage_window_width_interval.watch([this] { (void)reset(); });
+    _usage_disk_persistance_interval.watch([this] { (void)reset(); });
+}
+
+ss::future<> usage_manager::reset() {
+    if (
+      !bg_future_gate.is_closed()
+      && (ss::this_shard_id() == usage_manager_main_shard)) {
+        auto h = bg_future_gate.hold();
+        /// Calling start() will send any new configuration updates to the
+        /// newly initialized accounting fiber
+        if (_accounting_fiber) {
+            auto fiber = std::move(_accounting_fiber);
+            _accounting_fiber = nullptr;
+            co_await fiber->stop().finally([h] {});
```

Comment:
could you avoid this complexity and the need for the bg gate etc... by not resetting the background fiber, but rather resetting it's state?

Reply:
yeah nice idea , i think accounting_fiber::reset will remove the need for the bg gate

Reply:
i now remember why i did this. I wanted to have the usage manager allocate a much smaller amount of memory in the case the feature is disabled